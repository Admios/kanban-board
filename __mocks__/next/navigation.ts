const push = jest.fn();
export const useRouter = () => ({ push });
export const redirect = jest.fn();
export const navigation = { push, redirect };
