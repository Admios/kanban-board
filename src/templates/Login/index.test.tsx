import {
  startAuthentication,
  startRegistration,
} from "@simplewebauthn/browser";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Login } from ".";
import {
  generateAuthenticationOptions,
  generateRegistrationOptions,
  verifyAuthentication,
  verifyRegistration,
} from "./serverActions";
import { useRouter } from "next/navigation";

jest.mock("./serverActions.ts");
jest.mock("@simplewebauthn/browser");
const { push } = (useRouter as jest.Mock)();

afterEach(() => {
  jest.clearAllMocks();
});

it("should render", () => {
  const { container } = render(<Login />);
  expect(container).toMatchSnapshot("Default Login Page");
});

it("should login", async () => {
  render(<Login />);

  const email = "lorem@ipsum.com";
  const textBox = screen.getByRole("textbox", { name: "Email" });
  await userEvent.type(textBox, email);

  const loginButton = screen.getByRole("button", {
    name: "Login (Existing User)",
  });
  await userEvent.click(loginButton);

  expect(generateAuthenticationOptions).toHaveBeenCalledWith(email);
  expect(startAuthentication).toHaveBeenCalled();
  expect(verifyAuthentication).toHaveBeenCalled();
  expect(push).toHaveBeenCalledWith("/");
});

it("should login", async () => {
  render(<Login />);

  const email = "lorem@ipsum.com";
  const textBox = screen.getByRole("textbox", { name: "Email" });
  await userEvent.type(textBox, email);

  const signupButton = screen.getByRole("button", {
    name: "Register (New User)",
  });
  await userEvent.click(signupButton);

  expect(generateRegistrationOptions).toHaveBeenCalledWith(email);
  expect(startRegistration).toHaveBeenCalled();
  expect(verifyRegistration).toHaveBeenCalled();
  expect(push).toHaveBeenCalledWith("/");
});

it("should not login with empty email", async () => {
  render(<Login />);

  const loginButton = screen.getByRole("button", {
    name: "Login (Existing User)",
  });
  await userEvent.click(loginButton);

  expect(generateAuthenticationOptions).not.toHaveBeenCalled();
  expect(startAuthentication).not.toHaveBeenCalled();
  expect(verifyAuthentication).not.toHaveBeenCalled();
  expect(push).not.toHaveBeenCalled();
});

it("should not login with invalid email", async () => {
  render(<Login />);

  const email = "lorem";
  const textBox = screen.getByRole("textbox", { name: "Email" });
  await userEvent.type(textBox, email);

  const loginButton = screen.getByRole("button", {
    name: "Login (Existing User)",
  });
  await userEvent.click(loginButton);

  expect(generateAuthenticationOptions).not.toHaveBeenCalled();
  expect(startAuthentication).not.toHaveBeenCalled();
  expect(verifyAuthentication).not.toHaveBeenCalled();
  expect(push).not.toHaveBeenCalled();

  expect(screen.getByRole("alert")).toMatchSnapshot("Invalid Email Error");
});
