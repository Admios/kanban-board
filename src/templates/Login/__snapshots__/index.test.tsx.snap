// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not login with invalid email: Invalid Email Error 1`] = `
<div
  class="chakra-alert css-vgrb45"
  data-status="error"
  role="alert"
>
  <span
    class="chakra-alert__icon css-14ogjxt"
    data-status="error"
  >
    <svg
      class="chakra-icon css-qk6lof"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M11.983,0a12.206,12.206,0,0,0-8.51,3.653A11.8,11.8,0,0,0,0,12.207,11.779,11.779,0,0,0,11.8,24h.214A12.111,12.111,0,0,0,24,11.791h0A11.766,11.766,0,0,0,11.983,0ZM10.5,16.542a1.476,1.476,0,0,1,1.449-1.53h.027a1.527,1.527,0,0,1,1.523,1.47,1.475,1.475,0,0,1-1.449,1.53h-.027A1.529,1.529,0,0,1,10.5,16.542ZM11,12.5v-6a1,1,0,0,1,2,0v6a1,1,0,1,1-2,0Z"
        fill="currentColor"
      />
    </svg>
  </span>
  <div
    class="chakra-alert__title css-tidvy5"
    data-status="error"
  >
    There was an error
  </div>
  <div
    class="chakra-alert__desc css-zzks76"
    data-status="error"
  >
    Email is not valid
  </div>
</div>
`;

exports[`should render: Default Login Page 1`] = `
<div>
  <div
    class="css-lt6aye"
  >
    <div
      class="chakra-card css-igkfxv"
    >
      <div
        class="chakra-card__header css-1sl53ol"
      >
        <h2
          class="chakra-heading css-1dklj6k"
        >
          Login or Register
        </h2>
      </div>
      <div
        class="chakra-card__body css-1y4uwij"
      >
        <label
          class="chakra-form__label css-qxwrk0"
          for="email"
        >
          Email
        </label>
        <input
          class="chakra-input css-1ndqqtl"
          id="email"
          name="email"
          type="email"
          value=""
        />
      </div>
      <div
        class="chakra-card__footer css-t6dpxc"
      >
        <button
          class="chakra-button css-r7xd4a"
          type="button"
        >
          Authorize Me
        </button>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
